File: aoc_2023_ts/day04/day04.ts
import { parse } from "./parser";
import { readFileSync } from "fs";
class Card {
  id: number;
  have: number[];
  winning: number[];

  constructor(id: number, have: number[], winning: number[]) {
    this.id = id;
    this.have = have;
    this.winning = winning;
  }

  countWinningNumbers(): number {
    return this.have.filter((x) => this.winning.includes(x)).length;
  }

  getDayAValue(): number {
    let winningNumbers = this.countWinningNumbers();
    return winningNumbers == 0 ? 0 : 2 ** (winningNumbers - 1);
  }
}

function getCards(data: string): Card[] {
  return parse(data);
}

function dayA(filepath: string): number {
  let data = readFileSync(filepath).toString();
  let cards = getCards(data);
  return cards.map((card) => card.getDayAValue()).reduce((a, b) => a + b);
}
function dayB(filepath: string): number {
  let data = readFileSync(filepath).toString();
  let cards = getCards(data);
  let cardCounts: { [id: number]: number } = {};
  for (let card of cards) {
    cardCounts[card.id] = 1;
  }
  for (let i = 0; i < cards.length; i++) {
    for (let j = 0; j < cards[i].countWinningNumbers(); j++) {
      cardCounts[cards[i].id + j + 1] += cardCounts[cards[i].id];
    }
  }
  return Object.values(cardCounts).reduce((a, b) => a + b);
}

export { dayA, dayB, Card, getCards };

File: aoc_2023_py/day04/day04.py
class Card:
    def __init__(self, id: int, have: list[int], winning: list[int]):
        self.id = id
        self.have = have
        self.winning = winning

    @staticmethod
    def from_line(line: str) -> "Card":
        card_and_id, numbers = line.split(":")
        id = int(card_and_id[5:])
        have_str, winning_str = numbers.split("|")
        have = [int(number) for number in have_str.strip().split(" ") if number != ""]
        winning = [
            int(number) for number in winning_str.strip().split(" ") if number != ""
        ]
        return Card(id, have, winning)

    def __eq__(self, other: object) -> bool:
        return (
            isinstance(other, Card)
            and self.id == other.id
            and self.have == other.have
            and self.winning == other.winning
        )

    def __repr__(self):
        return (
            f"Card {self.id}: "
            f"{' '.join([str(number) for number in self.have])} "
            f"| {' '.join([str(number) for number in self.winning])}"
        )

    def calculate_matching(self) -> int:
        return len(set(self.have).intersection(set(self.winning)))

    def calculate_day_a_score(self) -> int:
        matching: int = self.calculate_matching()
        return 0 if matching == 0 else 2 ** (matching - 1)


def parse_lines(lines: list[str]) -> list[Card]:
    return [Card.from_line(line) for line in lines if line != ""]


def day_a(filepath: str) -> int:
    with open(filepath, "r") as f:
        lines: list[str] = f.readlines()
    return sum(card.calculate_day_a_score() for card in parse_lines(lines))


def day_b(filepath: str) -> int:
    with open(filepath, "r") as f:
        lines: list[str] = f.readlines()
    ticket_counts: dict[int, int] = {}
    cards: list[Card] = parse_lines(lines)
    for card in cards:
        ticket_counts.setdefault(card.id, 1)
        for i in range(card.calculate_matching()):
            ticket_counts.setdefault(card.id + i + 1, 1)
            ticket_counts[card.id + i + 1] += ticket_counts[card.id]
    return sum(ticket_counts.values())


def main():
    print(f"Day04 part a result is {day_a('data/input_data.txt')}")
    print(f"Day04 part b result is {day_b('data/input_data.txt')}")


if __name__ == "__main__":
    main()
